src/hotspot/cpu/x86/stubGenerator_x86_64.cpp

```diff
@@ -1922,6 +1922,9 @@ class StubGenerator: public StubCodeGenerator {
     BasicType type = is_oop ? T_OBJECT : T_INT;
     BarrierSetAssembler *bs = BarrierSet::barrier_set()->barrier_set_assembler();
     bs->arraycopy_prologue(_masm, decorators, type, from, to, count);
+    if (bs->enable_oop_arraycopy_prologue()) {
+      if (is_oop) bs->oop_arraycopy_prologue(_masm, decorators, type, from, to, count, rcx);
```

Comment:
`rcx` is `dst_obj`.

These aged x86 assembly code is written in a really poor way. `dword_count ` in L1895 is not an argument. The value of `dword_count` is actually loaded at L1930. The only arguments are `from`, `to`, and `count`. So I added the 4th argument (`rcx`) as dst_obj.

Reply:
With your explanation, I can see that they move `count` to `rcx` at L1930 and use `dword_count` to refer to it. But I still don't understand why `rcx` holds `dst_obj` here. Where is `rcx` set to `dst_obj`?

Reply:
There're two callers: [C2 arraycopy](https://github.com/mmtk/openjdk/blob/fedbddd89eb231767a854e9664af6c91478f8504/src/hotspot/share/opto/macroArrayCopy.cpp#L1046) and [C1 arraycopy](https://github.com/mmtk/openjdk/blob/fedbddd89eb231767a854e9664af6c91478f8504/src/hotspot/cpu/x86/c1_LIRAssembler_x86.cpp#L3494). The caller passes `dst_obj` to `rcx`.

The assembler generated follows system-v calling convention. c_rarg0~c_rarg5 means rdi, rsi, rdx, rcx, r8, r9. So as long as the caller passes a value as the 4th argument, it will be loaded to `rcx`.

Reply:
Well I'm also thinking about removing all the `oop_arraycopy_prologue`  code and use an array-range modbuf like jikesrvm [1]. This looks less hacky and more efficient. I'll discuss this with Steve.

[1] https://github.com/JikesRVM/JikesRVM/blob/master/MMTk/src/org/mmtk/plan/generational/GenMutator.java#L219-L222

Reply:
> There're two callers: [C2 arraycopy](https://github.com/mmtk/openjdk/blob/fedbddd89eb231767a854e9664af6c91478f8504/src/hotspot/share/opto/macroArrayCopy.cpp#L1046) and [C1 arraycopy](https://github.com/mmtk/openjdk/blob/fedbddd89eb231767a854e9664af6c91478f8504/src/hotspot/cpu/x86/c1_LIRAssembler_x86.cpp#L3494). The caller passes `dst_obj` to `rcx`.
> 
> The assembler generated functions follow system-v calling convention. c_rarg0~c_rarg5 means rdi, rsi, rdx, rcx, r8, r9. So as long as the caller passes a value as the 4th argument, it will be loaded to `rcx`.

I see. Thanks for the explanation.